CRM/Civicase/Event/Listener/CaseRoleCreation.php

```diff
@@ -67,12 +67,62 @@ public static function onPrepare(PrepareEvent $event) {
       return;
     }
 
-    $action = !empty($apiRequest['params']['id']) ? 'update' : 'create';
-    if ($action == 'create') {
+    $relationshipId = self::getRelationshipIdFromRequestParams($apiRequest['params']);
+    if ($relationshipId === NULL) {
       $caseCreationPreProcess = new CaseRoleCreationPreProcess();
       $caseCreationPreProcess->onCreate($apiRequest);
       $event->setApiRequest($apiRequest);
     }
+    elseif (!isset($apiRequest['params']['id'])) {
+      $apiRequest['params']['id'] = $relationshipId;
+      $event->setApiRequest($apiRequest);
+    }
+  }
+
+  /**
+   * Return the Id of the relationship for the API request params.
+   *
+   * The code looks from the ID param, or try to find it by the other received
+   * information.
+   *
+   * @param array $params
+   *   Params received from the API request.
+   *
+   * @return int|null
+   *   The ID of the relationship, or NULL in case of not being found.
+   */
+  private static function getRelationshipIdFromRequestParams(array $params) {
+    if (isset($params['id']) && is_numeric($params['id'])) {
+      return $params['id'];
+    }
+
```

Comment:
Can we add this code here?
```php
   $isSingleCaseRole = $this->isSingleCaseRole && !$this->isMultiClient;
    if (!$isSingleCaseRole) {
      return NULL;
    }
```
The reason is because block of code below is only useful for webforms when the single case role setting is turned on. If we leave it here, it means the logic to get the relationship ID will be run all the time even when not necessary when this is invoked from civicase.

Reply:
I can't put this code here, because those variables would not be defined in this class. 
Anyway, I was meditating about the situation of multiclient, and this logic seems to fit to me.
A relationship type, in a case, for two given people, can:
- Do not exist. The method will return NULL.
- Do exist, but it is not active (it has is_active=0 and an end_date). The method will return NULL too, because we want to be able to reassign a previously "deleted" relationship.
- Do exist, and it is active. The method returns the id of the relationship, because it can't be added again, no matter if it is multiple client or not, it is the same two contacts on the same case on the same relationship type. In which case it would have sense to have that same exact row on relationship? 

Let me know if there is something that I am misunderstanding. If you think that is still needed to check multiple client and all that stuff, I can try to find a way to load that information here, or coordinate on another place the validation.